lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp

```diff
@@ -0,0 +1,178 @@
+
+    // TODO: when supported, iterate over the tensor of sizes. This will be
+    // iterating through a level of indirection.
+
+    // Construct the loop bounds based on the canonical arithmetic progression.
+    Location loc = tileOp.getLoc();
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value totalSize =
+        rewriter.create<tensor::DimOp>(loc, tileOp.outs().front(), zero);
+    Value step = tileOp.tile_sizes();
+    assert(step.getType().isa<IndexType>() && "NYI: not an index type");
+
+    AffineBuilder ab(rewriter, loc);
+    AffineExpr i, j, M;
+    bindDims(rewriter.getContext(), i, j);
+    bindSymbols(rewriter.getContext(), M);
+    Value sizePerThread = ab.ceil(Tie{i, totalSize}, Tie{M, step});
+
+    // Construct the op without a body builder: we need to clone the ops in the
+    // body explicitly after having access to the new bbArgs.
+    // As a consequence, `ensureTerminator` is not called and the body has no
+    // terminator.
+    linalg_ext::InParallelOp inParallelOp =
+        rewriter.create<linalg_ext::InParallelOp>(loc, tileOp->getResultTypes(),
+                                                  sizePerThread);
+    inParallelOp.ensureTerminator(inParallelOp.region(), rewriter, loc);
+
+    // At the beginning of the InParallelOp, compute offset and sizes.
+    rewriter.setInsertionPointToStart(inParallelOp.getBody());
+
+    // Materialize the implicit subtensors as explicit subset_extract.
+    // TODO: generalize to multiple offset/chunk_size bbargs if needed.
+    // TODO: generalize the subset op.
+    Value offset = ab.mul(Tie{i, inParallelOp.getThreadIndex()}, Tie{M, step});
+    Value size = ab.min({ab.sub(Tie{i, totalSize}, Tie{j, offset}), step});
+
+    SmallVector<Value> implicitSubtensorExtracts;
+    for (Value tensor : tileOp.outs()) {
+      implicitSubtensorExtracts.push_back(
+          rewriter.createOrFold<tensor::ExtractSliceOp>(loc, tensor, offset,
+                                                        size, one));
+    }
+
+    // Get a reference to the TileOp terminator before the body is merged and it
+    // becomes too hard to get to the terminator.
+    auto tileYieldOp = cast<TileYieldOp>(tileOp.getBody()->getTerminator());
+
+    // Regroup the values that replace the tileOp's bbArg and move the body.
+    SmallVector<Value> bbArgsTranslated{offset, size};
+    llvm::append_range(bbArgsTranslated, implicitSubtensorExtracts);
+    rewriter.mergeBlockBefore(&tileOp.region().front(),
+                              inParallelOp.getBody()->getTerminator(),
+                              bbArgsTranslated);
+
+    // tileOp's terminator is not the terminator, insert explicit subset_insert
+    // ops and feed them to a new scf.yield terminator that we can now add.
+    auto performConcurrentlyOp =
+        cast<PerformConcurrentlyOp>(&inParallelOp.getBody()->back());
```

Comment:
Should we add a helper to `InParallelOp` that returns the terminator already cast to `PerformConcurrentlyOp`? We do that in a whole bunch of places